README.md

# PopSwitch

A switch that has a fun spring animation added.

Works for iOS 10+
Note: it is much faster than this gif shows.

![](https://github.com/wvabrinskas/PopSwitch/blob/master/public/PopSwitch.gif)

# Installation
* Simply install using Cocoapods, add `'pod PopSwitch'` to your podfile

# Simple to use
* `import PopSwitch`
* Create a SwitchColor for your switch: `let color:SwitchColor = (background: UIColor.white.cgColor, switch: UIColor.green.cgColor)`
* Create an instance of PopSwitch `let popSwitch = PopSwitch(position: .Off, color: color, type: .Switch)`
  * `color` can be `nil`. Will default to system white and green.
  * There are two positions: `PopSwitch.State.On` or `PopSwitch.State.Off`
  * There are two switch types you can pick from:
    * `SwitchType.Radio`
    * `SwitchType.Switch`
* Set the `PopSwitchDelegate` for your PopSwitch
  * `popSwitch.delegate = self`
* Conform to the delegate function `valueChanged(to state: PopSwitch.State)`
* Add your PopSwitch to a view `view.addSubview(popSwitch)`


CHANGELOG.md

# CHANGELOG

# 1.0.10
* Added the ability to generate a `PopSwitch` of two types (`SwitchType` enum): `Radio` and `Switch`.


PopSwitch/AppDelegate.swift
//
//  AppDelegate.swift
//  PopSwitch
//
//  Created by William Vabrinskas on 7/27/17.
//  Copyright © 2017 williamvabrinskas. All rights reserved.
//

import UIKit

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {

    var window: UIWindow?


    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
        // Override point for customization after application launch.
        return true
    }

    func applicationWillResignActive(_ application: UIApplication) {
        // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
        // Use this method to pause ongoing tasks, disable timers, and invalidate graphics rendering callbacks. Games should use this method to pause the game.
    }

    func applicationDidEnterBackground(_ application: UIApplication) {
        // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
        // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
    }

    func applicationWillEnterForeground(_ application: UIApplication) {
        // Called as part of the transition from the background to the active state; here you can undo many of the changes made on entering the background.
    }

    func applicationDidBecomeActive(_ application: UIApplication) {
        // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
    }

    func applicationWillTerminate(_ application: UIApplication) {
        // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
    }


}



PopSwitch-AppDemo/PopSwitch/ViewController.swift
//
//  ViewController.swift
//  PopSwitch
//
//  Created by William Vabrinskas on 7/27/17.
//  Copyright © 2017 williamvabrinskas. All rights reserved.
//

import UIKit


enum Color: Int {
    case LightGreen,DarkGreen,ClearDarkGreen,Purple,ErrorRed,LightErrorRed,LightSuccessGreen,
    LightPurple,LightTabBarBlue,DarkPurple,DarkBlue,LightBlue
    
}

protocol ColorDefinition {
    func uiColor() -> UIColor
}

enum ColorModel: ColorDefinition {
    case Color(color:Color)
    
    func uiColor() -> UIColor {
        switch self {
        case let .Color(color):
            switch color {
            case .LightGreen:
                return UIColor(red: 1/255, green: 158/255, blue: 134/255, alpha: 1)
            case .DarkGreen:
                return UIColor(red: 0/255, green: 106/255, blue: 85/255, alpha: 1)
            case .ClearDarkGreen:
                return UIColor(red: 0/255, green: 106/255, blue: 85/255, alpha: 0.6)
            case .Purple:
                return UIColor(red: 117.0/255.0, green: 32.0/255.0, blue: 167.0/255.0, alpha: 1.0)
            case .DarkPurple:
                return UIColor(red: 127.0/255.0, green: 68.0/255.0, blue: 198.0/255.0, alpha: 1.0)
            case .ErrorRed:
                return UIColor(red: 214.0/255.0, green: 41.0/255.0, blue: 44.0/255.0, alpha: 1.0)
            case .LightErrorRed:
                return UIColor(red: 239.0/255.0, green: 119.0/255.0, blue: 124.0/255.0, alpha: 1.0)
            case .LightSuccessGreen:
                return UIColor(red: 106.0/255.0, green: 200.0/255.0, blue: 95.0/255.0, alpha: 1.0)
            case .LightPurple:
                return UIColor(red: 94.0/255.0, green: 101.0/255.0, blue: 168.0/255.0, alpha: 1.0)
            case .LightTabBarBlue:
                return UIColor(red: 193.0/255.0, green: 211.0/255.0, blue: 243.0/255.0, alpha: 1.0)
            case .DarkBlue:
                return UIColor(red: 36.0/255.0, green: 88.0/255.0, blue: 165.0/255.0, alpha: 1.0)
            case .LightBlue:
                return UIColor(red: 62.0/255.0, green: 109.0/255.0, blue: 179.0/255.0, alpha: 1.0)
            }
        }
    }
}


class ViewController: UIViewController {
    var popSwitch: PopSwitch!
    @IBOutlet weak var centerView: UIView!

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        let color:SwitchColor = (background:ColorModel.Color(color: .LightGreen).uiColor().cgColor, switch: ColorModel.Color(color: .LightTabBarBlue).uiColor().cgColor)
        popSwitch = PopSwitch(position: .Off, color: color, type: .Switch)
        centerView.addSubview(popSwitch)
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    

}



PopSwitch/PopSwitch.swift
//
//  PopSwitch.swift
//  PopSwitch
//
//  Created by William Vabrinskas on 7/27/17.
//  Copyright © 2017 williamvabrinskas. All rights reserved.
//

import Foundation
import UIKit
import CoreGraphics
import QuartzCore

public typealias SwitchColor = (background: CGColor?, switch: CGColor?)

open class PopSwitch: UIView {
    
    public enum State {
        case On,Off
    }
    
    public enum SwitchType {
        case Radio, Switch
    }
    
    open var state:State!
    open var delegate:PopSwitchDelegate?
    
    private var color:SwitchColor?
    private static let height:CGFloat = 30.0
    private var switchType: SwitchType!
    
    private lazy var startOnXOrigin:CGFloat = {
        return self.frame.width - self.circle.frame.size.width
    }()
    private lazy var onXOrigin: CGFloat = {
        return self.frame.width - (self.circle.frame.size.width / 2)
    }()
    
    private lazy var offXOrigin: CGFloat = {
        return self.circle.frame.size.width / 2
    }()
    
    private lazy var switchLayer:CAShapeLayer = {
        let shapeLayer = CAShapeLayer()
        shapeLayer.frame = self.bounds
        shapeLayer.path = UIBezierPath.init(roundedRect: self.bounds, cornerRadius: PopSwitch.height).cgPath
        
        if self.switchType == .Switch {
            if self.state == .On {
                shapeLayer.fillColor = self.color?.background ?? UIColor.white.cgColor
                self.circle.frame.origin = CGPoint(x: self.startOnXOrigin, y: 0)
            } else {
                shapeLayer.fillColor = self.getDarkerColor()
                self.circle.frame.origin = CGPoint(x: 0, y: 0)
            }
        } else {
            shapeLayer.fillColor = self.getDarkerColor()
            shapeLayer.lineWidth = 5.0
            shapeLayer.strokeColor = self.color?.background ?? UIColor.white.cgColor
        }
        
        shapeLayer.addSublayer(self.circle)
        
        return shapeLayer
    }()
    
    private lazy var circle:CAShapeLayer = {
        let circleLayer = CAShapeLayer()
        var height = PopSwitch.height
        var center = CGPoint(x:0, y:0)
        
        if self.switchType == .Radio {
            height = PopSwitch.height - 7
            center = CGPoint(x: (PopSwitch.height / 2) - (height / 2), y: (PopSwitch.height / 2) - (height / 2))
        }
        
        circleLayer.frame = CGRect(x: center.x, y: center.y, width: height, height: height)
        circleLayer.path = UIBezierPath.init(ovalIn: circleLayer.bounds).cgPath
        circleLayer.fillColor = self.color?.switch ?? UIColor.green.cgColor
        return circleLayer
    }()
    
    fileprivate func getDarkerColor() -> CGColor {
        var newColor:CGColor?
        
        if let components = color?.background?.components {
            if components.count >= 3 {
                let r = (components[0] * 255) - 100
                let g = (components[1] * 255) - 100
                let b = (components[2] * 255) - 100
                
                newColor = UIColor(red: r/255.0, green: g/255.0, blue: b/255.0, alpha: 1.0).cgColor
            }
        }
        
        return newColor ?? color?.background ?? UIColor.white.cgColor
    }
    
    fileprivate func onSpringAnimation() -> CASpringAnimation {
        let spring = CASpringAnimation(keyPath: "position")
        spring.damping = 10.0
        spring.duration = spring.settlingDuration
        spring.repeatCount = 0
        spring.speed = 2
        spring.toValue = [onXOrigin, self.circle.frame.size.height / 2]
        spring.initialVelocity = 0
        spring.fillMode = kCAFillModeBoth
        spring.isRemovedOnCompletion = false
        return spring
    }
    
    fileprivate func offSpringAnimation() -> CASpringAnimation {
        let spring = CASpringAnimation(keyPath: "position")
        spring.damping = 10.0
        spring.speed = 2
        spring.duration = spring.settlingDuration
        spring.repeatCount = 0
        spring.toValue = [offXOrigin, self.circle.frame.size.height / 2]
        spring.initialVelocity = 0
        spring.fillMode = kCAFillModeBoth
        spring.isRemovedOnCompletion = false
        return spring
    }
    
    fileprivate func scaleDownAnimation() -> CASpringAnimation {
        let spring = CASpringAnimation(keyPath: "transform.scale")
        spring.damping = 10.0
        spring.speed = 2
        spring.duration = spring.settlingDuration
        spring.repeatCount = 0
        spring.toValue = [0.0,0.0]
        if switchType == .Switch {
            spring.toValue = [0.6,0.6]
        }
        spring.initialVelocity = 0
        spring.fillMode = kCAFillModeBoth
        spring.isRemovedOnCompletion = false
        return spring
    }
    
    fileprivate func scaleUpAnimation() -> CASpringAnimation {
        let spring = CASpringAnimation(keyPath: "transform.scale")
        spring.damping = 10.0
        spring.speed = 2
        spring.duration = spring.settlingDuration
        spring.repeatCount = 0
        spring.toValue = [1.0,1.0]
        spring.initialVelocity = 0
        spring.fillMode = kCAFillModeBoth
        spring.isRemovedOnCompletion = false
        return spring
    }
    
    fileprivate func colorOnChangeAnimation() -> CABasicAnimation {
        //backgroundColor
        let colorChange = CABasicAnimation(keyPath: "fillColor")
        colorChange.toValue = self.color?.background ?? UIColor.white.cgColor
        colorChange.isRemovedOnCompletion = false
        colorChange.fillMode = kCAFillModeBoth
        colorChange.repeatCount = 0
        return colorChange
    }
    
    fileprivate func colorOffChangeAnimation() -> CABasicAnimation {
        //backgroundColor
        let colorChange = CABasicAnimation(keyPath: "fillColor")
        colorChange.toValue = getDarkerColor()
        colorChange.repeatCount = 0
        colorChange.fillMode = kCAFillModeBoth
        colorChange.isRemovedOnCompletion = false
        return colorChange
    }
    
    private func animate(to state:State) {
        
        if switchType == .Switch {
            let scaleGroup = CAAnimationGroup()
            scaleGroup.animations = [self.scaleDownAnimation(), self.scaleUpAnimation()]
            circle.add(scaleGroup, forKey: "scale")
            
            if state == .On {
                //animating to the ON position
                circle.add(self.onSpringAnimation(), forKey: "onAnimation")
                switchLayer.add(self.colorOnChangeAnimation(), forKey: "onColorAnimation")
            } else {
                circle.add(self.offSpringAnimation(), forKey: "offAnimation")
                switchLayer.add(self.colorOffChangeAnimation(), forKey: "offColorAnimation")
            }
        } else {
            if state == .On {
                circle.add(self.scaleUpAnimation(), forKey: "onAnimation")
            } else {
                circle.add(self.scaleDownAnimation(), forKey: "offAnimation")
            }
        }
        
    }
    
    public init(position state:State, color: SwitchColor?, type:SwitchType) {
        let width = type == .Radio ? PopSwitch.height : PopSwitch.height * 1.8
        
        super.init(frame: CGRect(x: 0, y: 0, width: width, height: PopSwitch.height))
        
        self.switchType = type
        
        self.backgroundColor = .clear
        self.state = state
        self.color = color
        self.layer.addSublayer(switchLayer)
        
        let touchGesture = UITapGestureRecognizer(target: self, action: #selector(changeState))
        touchGesture.numberOfTapsRequired = 1
        touchGesture.numberOfTouchesRequired = 1
        self.addGestureRecognizer(touchGesture)
        
        setState(state: state, callback: false)
    }
    
    //through touch gesture
    @objc private func changeState() {
        if self.state == .On {
            self.state = .Off
        } else {
            self.state = .On
        }
        animate(to: self.state)
        
        delegate?.valueChanged(control: self)
    }
    
    //programmatically set state
    open func setState(state: State, callback:Bool) {
        self.state = state
        animate(to: state)
        if callback {
            delegate?.valueChanged(control: self)
        }
    }
    
    required public init?(coder aDecoder: NSCoder) {
        fatalError("Cannot add PopSwitch as part of interface builder. Sorry =(")
    }
}

